### PR TITLE
arrays-referencing_array_and_its_elements-medium-q10

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/arrays/referencing_array_and_its_elements/medium/q10/VowelNames.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package arrays.referencing_array_and_its_elements.medium.q10;
+
+/*
+Write a Java program that takes a String array of 6 given names. Then, count how many of those names begin with the
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/VowelNames.java:4]: Change question text to:
"Write a Java program that asks the user for 6 names and stores them in an array of Strings. Then, count how many of the names start with the letter "B" and display the value.

Sample Output:
Enter 6 given names:
Billy Bob Jacob Anthony Bartholomew Angus
There are 3 names that start with the letter B."

File: src/arrays/referencing_array_and_its_elements/medium/q10/VowelNames.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        String[] arr = new String[6];
+        System.out.println("Enter 6 given names:");
+        for (int i = 0; i < 6; i++)
+            arr[i] = in.next();
+        int count = 0;
+        for (int i = 0; i < arr.length; i++) {
+            if (arr[i].charAt(0) == 'B')
+                count++;
+        }
+        System.out.println("There are " + count + " names start with the letter B");
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/VowelNames.java:28]:
```suggestion
        System.out.println("There are " + count + " names that start with the letter B");
```

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+public class MainTest extends BaseTest {
+    // Java
+
+    public static int n = 6;
+    public static char c = 'B';
+
+    public Clause[] testSentence() {
+        TestOption.isInputTest = true;
+        TestOption.defaultInput = arrayToInput(replicateArray('A', n));
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:25]: switch to ArrayUtil's generate random words method.

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        return new Clause[]{
+                new StringLiteral("Enter " + n + " given names:"),
+                new Optional(new StringLiteral(" ")),
+                new NewLine(),
+                new StringLiteral("There are "),
+                new IntegerLiteral("ans"),
+                new StringLiteral(" names start with the letter B"),
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:32]:
```suggestion
                new StringLiteral(" names that start with the letter B"),
```

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            String st = "" + (char) ('A' + r.nextInt(26));
+            for (int j = 1; j < up; j++)
+                st += "" + (char) ('a' + r.nextInt(26));
+            ans += st + " ";
+        }
+
+        return ans.trim();
+    }
+
+    static Stream<String> InputProvider() {
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:55]: Change to using ArrayUtil generate random word array

Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:55]:
```suggestion
    static Stream<String> inputProvider() {
```

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                randStr(),
+                randStr(),
+                arrayToInput(replicateArray(c, n)),
+                arrayToInput(replicateArray((char) (c + 1), n))
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("InputProvider")
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:70]:
```suggestion
    @MethodSource("inputProvider")
```

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    void printsCorrectOutput(String a) {
+        int t = 0;
+        for (int i = 0; i < a.length(); i++)
+            if (a.charAt(i) == c)
+                t++;
+        runWithInput(a);
+        assertEquals(t, Integer.parseInt(getItemByName("ans")));
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:77]: add error message

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java (hunk 5 above)
Comment (SamamthaH) [reply src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java]: If you don't think a change should be implemented you must explain why.

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java (hunk 7 above)
Comment (SamamthaH) [reply src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:77]: You did not add an error message.

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/medium/q10/VowelNames.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package arrays.referencing_array_and_its_elements.medium.q10;
+
+/*
+Write a Java program that asks the user for 6 names and stores them in an array of Strings. Then, count how many of the
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/VowelNames.java:4]: Change this so it displays the names that start with B not just counting them, and increasing the input size.

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            for (int j = 1; j < up; j++)
+                st += "" + (char) ('a' + r.nextInt(26));
+            ans += st + " ";
+        }
+
+        return ans;
+    }
+
+    static Stream<String> inputProvider() {
+        return Stream.of(
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:56]: why make an entirely new method instead of just using the random word generation from array util?

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java (hunk 9 above)
Comment (KaiKhademi) [reply src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:56]: I wanted the words to start with capital letters, like the question

File: src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            }
+            if(flg)
+                ans += st.charAt(i);
+            if(st.charAt(i) == ' ')
+                flg = false;
+        }
+        runWithInput(st, new Clause[]{
+                new StringLiteral(ans.trim())
+        });
+        CustomAssertions._assertEquals(t, Integer.parseInt(getItemByName("ans")), "Your program does not correctly print the number of names starting with " + ch + ".");
```
Comment (SamamthaH) [inline src/arrays/referencing_array_and_its_elements/medium/q10/MainTest.java:93]:
```suggestion
        CustomAssertions._assertEquals(t, Integer.parseInt(getItemByName("ans")), "Your program does not print the correct number of names starting with " + ch + ".");
```
